pulumi/src/index.ts

```diff
@@ -0,0 +1,323 @@
+                                `${bucketARN}/${completeName}/keystore.jks`,
+                            ],
+                        },
+                    ],
+                };
+                return JSON.stringify(policy);
+            });
+
+        const lambdaRolePolicy = new aws.iam.RolePolicy(`${baseName}-lambda-cert-function-policy`, {
+            role: lambdaRole,
+            policy: lambdaPolicy,
+        }, roleParent);
+
+        const certOpts = Object.assign({
+            dependsOn: [lambdaRolePolicy],
+        }, instanceParent);
+
+        const environment = all([keyBucket.bucket, kmsKey.keyId])
+            .apply(([bucketName, kmsKeyId]) => {
+                return {
+                    ACME_SERVER_URL: acmeUrl,
+                    KMS_KEY_ID: kmsKeyId,
+                    BUCKET_NAME: bucketName,
+                    CERTIFICATE_NAME: `${inputs.domainNamePrefix}.${inputs.route53DomainName}`,
+                    ADMIN_EMAIL: inputs.adminEmail,
+                    GENERATE_JAVA_KEYSTORE: inputs.generateJavaKeyStore ? "true" : "false",
+                };
+            });
+
+        const certTags = Object.assign({
+            Name: `Maintain TLS Keys for ${inputs.description}`,
+        }, inputs.baseTags);
+
+        const certFunction = new aws.lambda.Function(`${baseName}-lambda-cert-function`, {
+            description: `Maintain TLS Certificate for ${inputs.description}`,
+            runtime: aws.lambda.Go1dxRuntime,
+            code: new asset.RemoteArchive(releaseURL),
+            handler: "lambda-cert",
+            role: lambdaRole.arn,
+            timeout: 300,
+            environment: {
+                variables: environment,
+            },
+            tags: certTags,
+        }, certOpts);
+
+        const triggerOpts = Object.assign({
+            dependsOn: [certFunction],
+        }, instanceParent);
+        const trigger = new aws.cloudwatch.EventRule(`${baseName}-lambda-cert-schedule`, {
+            description: `Check ${inputs.description} TLS Cert Status`,
+            scheduleExpression: "rate(12 hours)",
+        }, triggerOpts);
+
+        const triggerParent = {parent: trigger};
+        new aws.cloudwatch.EventTarget(`${baseName}-lambda-cert-target`, {
+            rule: trigger.name,
+            targetId: certFunction.name,
+            arn: certFunction.arn,
+        }, triggerParent);
+
+        const lambdaParent = {parent: certFunction};
+        new aws.lambda.Permission(`${baseName}-lambda-cert-permission`, {
+            statementId: "AllowExecutionFromCloudWatch",
+            action: "lambda:InvokeFunction",
+            function: certFunction,
+            principal: "events.amazonaws.com",
+            sourceArn: trigger.arn,
+        }, lambdaParent);
+
+        return instance;
+    }
+
+    private constructor(name: string, opts?: ResourceOptions) {
+        super("operator-error:aws:LambdaCert", name, {}, opts);
```

Comment:
One thing to mention, which I am largely mentioning not because I think you should necessarily do it, but because I'd love feedback on this, is that components can have tracked inputs and outputs.

For example, instead of `{}`, I think you'd supply the `inputs` passed to `create`. And then you'd make a call to `this.registerResourceOutputs(...)` with the resulting `LambdaCertOutputs`.

This would cause the state to get recorded in the resulting checkpoint state. But the reason I'm questioning this whole area of our model is that I don't see how that'd immediately improve your life in any way here. (A minor point, it would allow us to show that information in the pulumi.com service UI ...)

Is that even remotely interesting in this case?

Reply:
Having the information shown in the service UI is definitely interesting. I'll look at adding this.